src/vm/proftoeeinterfaceimpl.cpp

```diff
@@ -9804,6 +9808,75 @@ HRESULT ProfToEEInterfaceImpl::ApplyMetaData(
     return hr;
 }
 
+// Add path to list of assemblies used by the runtime to probe assembly references
+// pAssemblyPath should be absolute path of assembly
+// Assembly name should match filename.
+HRESULT ProfToEEInterfaceImpl::AddAssemblyPath(AppDomainID appDomainId,
+    const WCHAR *pAssemblyPath)
+{
+    CONTRACTL
+    {
+        NOTHROW;
+        GC_NOTRIGGER;
+        MODE_ANY;
+        CANNOT_TAKE_LOCK;
+        SO_NOT_MAINLINE;
+    }
+    CONTRACTL_END;
+
+    PROFILER_TO_CLR_ENTRYPOINT_SYNC_EX(kP2EEAllowableAfterAttach, (LF_CORPROF, LL_INFO1000, "**PROF: AddAssemblyPath.\n"));
+
+    BaseDomain* pDomain = reinterpret_cast<BaseDomain*>(appDomainId);
+    if (pDomain == NULL)
+    {
+        return CORPROF_E_DATAINCOMPLETE;
+    }
+
+    HRESULT hr = S_OK;
+    EX_TRY
+    {
+        SString fileName(pAssemblyPath);
+        SString simpleName;
+        bool isNativeImage;
+
+        CLRPrivBinderCoreCLR* binderContext = pDomain->GetTPABinderContext();
+        if (binderContext == nullptr)
+        {
+            hr = CORPROF_E_NOT_YET_AVAILABLE;
```

Comment:
This is really problem, as AppDomain may either not have fusion configuration at all, or it may not been called. Per my tests, when AppDomainCreationFinished called, it will not be configured, same as on first calls of GetReJITParameters.
I'm also not sure what will be recommended way to get AppDomainId - should it be resolved from thread id before applying IL rewrites? As AppDomainCreationFinished was called 3 times - second domain was real application domain.

Reply:
A few things in here:
1) There are three different domains reported to maintain parity with desktop .NET runtime. They are the system domain, the shared domain, and the default domain. For .Net Core apps only the default domain is going to be relevant. The other two will never have a binding context or load assemblies.

2) It appears I was wrong about the timing of AppDomainCreationFinished. I thought it occurred later but it actually occurs really early, before the AppDomain has initialized any of the binder state.

3) If you are trying to ReJIT every managed method then GetReJITParameters will occur before the default AppDomain has been fully configured because managed code is used as part of the configuration process. For example this is the callstack when the ApplicationContext is first created and you can see two calls from System.Private.CoreLib.dll already in there. 

>	coreclr.dll!AppDomainNative::CreateContext(AppDomainBaseObject * refThisUNSAFE) Line 122	C++
 	System.Private.CoreLib.dll!00007ffcaf0b0db3()	Unknown
 	System.Private.CoreLib.dll!00007ffcaf0b1b77()	Unknown
 	coreclr.dll!CallDescrWorkerInternal() Line 101	Unknown
 	coreclr.dll!CallDescrWorkerWithHandler(CallDescrData * pCallDescrData, int fCriticalCall) Line 81	C++
 	coreclr.dll!MethodDescCallSite::CallTargetWorker(const unsigned __int64 * pArguments, unsigned __int64 * pReturnValue, int cbReturnValue) Line 632	C++
 	coreclr.dll!MethodDescCallSite::Call(const unsigned __int64 * pArguments) Line 472	C++
 	coreclr.dll!AppDomain::InitializeDomainContext(int allowRedirects, const wchar_t * pwszPath, const wchar_t * pwszConfig) Line 7710	C++
 	coreclr.dll!SystemDomain::InitializeDefaultDomain(int allowRedirects, ICLRPrivBinder * pBinder) Line 3198	C++
 	coreclr.dll!SystemDomain::SetupDefaultDomain() Line 3131	C++
 	coreclr.dll!SystemDomain::SetupDefaultDomainNoThrow() Line 3154	C++
 	coreclr.dll!EEStartup(tagCOINITEE fFlags) Line 1183	C++
 	coreclr.dll!EnsureEEStarted(tagCOINITEE flags) Line 357	C++
 	coreclr.dll!InitializeEE(tagCOINITEE flags) Line 295	C++
 	coreclr.dll!CorRuntimeHostBase::Start() Line 219	C++
 	coreclr.dll!CorHost2::Start() Line 177	C++

Perhaps in the future we could add an explicit event that says when the binder state has configured, but for now the easy option is probably for profilers to try to do it during ModuleLoadFinished, prior to adding the AssemblyRef to the assembly. If this is done for System.Private.CoreLib.dll it would fail (because CoreLib loads before binding paths are initialized) but for any other assembly it should work OK.